### DISCUSSION TITLE
OWASP ASVS: Checklist or Catalog of Controls

### DISCUSSION THREAD

**Discussion by ohsh6o** [usnistgov/OSCAL#869] [Q&A]
This has come up before, but there appears to be a growing need for showing the promise of OSCAL outside _just_ US governmental control sets. One discussed before in previous meetings is OWASP ASVS. For reference, community interest in this idea: OWASP/ASVS#817.

OSCAL aside, I can see some of this set of advisories as a tactical checklist. They themselves define individual items as controls.  [For reference, the example below is the final product of the current 4.0.2](https://raw.githubusercontent.com/OWASP/ASVS/v4.0.2/4.0/OWASP%20Application%20Security%20Verification%20Standard%204.0.2-en.pdf).

<img width="889" alt="Screen Shot 2021-03-17 at 10 14 38 PM" src="https://user-images.githubusercontent.com/61464190/111563339-2ca38b80-876e-11eb-9026-01979dff600a.png">

Given that OSCAL defines [controls in catalogs](https://pages.nist.gov/OSCAL/documentation/schema/catalog-layer/catalog/) to be more complex with parameters, objectives, and assessments, sub-controls, and references, does this warrant a separate model outside of catalogs of controls, and we make an issue for "checklist" or something analogous as a first-order concern in OSCAL?

I do not have a set opinion, and thought community feedback is warranted.

**Comment by GaryGapinski** [comment]
The ASVS can be cast as an OSCAL Catalog with an OSCAL Profile for each of its levels. I do not think it is necessary to create an additional OSCAL not-a-catalog element. An admittedly partial fidelity transform would be straightforward (using [this](https://github.com/OWASP/ASVS/blob/master/4.0/docs_en/OWASP%20Application%20Security%20Verification%20Standard%204.0.2-en.xml) as input).

The ASVS contains front matter (not _yet_ accommodated by OSCAL and worthy of an additional OSCAL element), groups of controls, controls, subordinate controls, and back matter (the ASVS back matter appears to be only partially accommodated by the OSCAL `<back-matter>` element: only Appendix B is a good fit). The ASVS prose can be used to construct control requirement statements, control objectives, control-specific parameters (e.g. [V3.3.2](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x12-V3-Session-management.md)), and assessment methods. There are intra-control references to other documents.

The ASVS [CONTRIBUTING.md](https://github.com/OWASP/ASVS/blob/master/CONTRIBUTING.md) mandates control identifier preservation in successive revisions. While that seems to be a popular fashion it is not universally practiced.

There is internationalization but it is not driven by a single set of proto-document fragments (the English and French markdown documents are separate, and additional translations have no apparent markdown counterparts).

**Comment by GaryGapinski** [comment]
[ASVStoOSCAP.zip](https://github.com/usnistgov/OSCAL/files/6173505/ASVStoOSCAP.zip)
(archived to appease a prissy attachment gate) will transform the ASVS XML to OSCAL Catalog and Profiles. It's rather bare bones: the XML is lacking as is the transform. The ASVS prose is essentially assessment imperatives which would require non-trivial manipulation to create corresponding objectives and statements.

**Reply by ohsh6o** [reply]
@GaryGapinski, you do not mess around! I will definitely want to review this in the next week or so, because I believe what is suggested there is to shift to an OSCAL-based pipeline to manage artifacts and make that the primary mode of publication, if they will indulge it, but your front matter comment and others made suggest that might be a tad ambitious. I will play around with this soon, thanks!
